infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/LogReplicationAckReader.java

```diff
@@ -98,28 +98,51 @@ private long getMaxReplicatedStreamsTail() {
      *
      * For Log Entry Sync, this function returns the total number of entries remaining to be sent across all replicated
      * streams.
-     * For Snapshot Sync, each entry sent is a snapshot.  So this function returns the total number of snapshots
-     * remaining to be sent.
-     * If the ack'd timestamp is uninitialized, it returns 100%, which means no replication has been done.
+     *
+     * For Snapshot Sync, each entry sent is a snapshot of the corresponding stream.  The stream could have been
+     * checkpointed and trimmed so we cannot find the remaining number of entries accurately.  In this case, we simply
+     * subtract the acked timestamp from the global log tail when the snapshot sync started.
+     * Note that this method is not accurate because the global tail advances on writes for both replicated and
+     * non-replicated streams.  So it does not accurately represent the remaining entries to send for replicated streams.
+     *
+     * If there is no data on the active, it returns 0, which means no replication remaining.
+     * If the ack'd timestamp is uninitialized(no ack received), it returns the log tail, which means no replication has
+     * been done.
      */
     private long calculateRemainingEntriesToSend(long ackedTimestamp) {
         long maxReplicatedStreamTail = getMaxReplicatedStreamsTail();
 
+        // No data to send on the Active, so no replication remaining
         if (maxReplicatedStreamTail == Address.NON_ADDRESS) {
-            // No data to send.  No Replication remaining
-            return NO_REPLICATION_REMAINING_PERCENT;
+            return NO_REPLICATION_REMAINING_ENTRIES;
         }
-        if (ackedTimestamp == Address.NON_ADDRESS) {
-            return FULL_REPLICATION_REMAINING_PERCENT;
+
+        // If doing a snapshot sync and nothing has been acked, all replication is remaining.
+        // So send the stream tail.
+        // NOTE: This is an approximation
+        if (ackedTimestamp == Address.NON_ADDRESS &&
+                lastSyncType == LogReplicationMetadata.ReplicationStatusVal.SyncType.SNAPSHOT) {
```

Comment:
I was thinking, we might not need this, right? Also it could be imprecise, in the case deltas are already coming in, so the maxReplicatedStreamTail is moving beyond global tail... we could just calculate all snapshot sync in L145, right?

Reply:
Yes, that's a good option.